src/sdl2_compat.c

```diff
@@ -5619,6 +5614,22 @@ static SDL_InitFlags PreInitSubsystem(SDL_InitFlags flags)
         }
 #endif
         SDL3_SetHint(SDL_HINT_IME_IMPLEMENTED_UI, hint);
+
+        /* Configure fractional scaling depending on the requested video driver
+         * TODO: Add quirks for this if the heuristic here isn't good enough */
+        hint = SDL3_GetHint("SDL_VIDEO_DRIVER");
+        if (!hint) {
+            hint = SDL3_GetHint("SDL_VIDEODRIVER");
+        }
+        if (!hint || SDL3_strcasecmp(hint, "wayland") != 0) {
+            /* If this app does not explicitly ask for Wayland, avoid using Wayland fractional scaling.
+             * This is more compatible with applications that have only been tested under X11 without high DPI support.
+             * Full discussion is here: https://github.com/libsdl-org/SDL/issues/12158 */
+            SDL3_SetHint(SDL_HINT_VIDEO_WAYLAND_SCALE_TO_DISPLAY, "1");
+        } else {
+            /* If the app did ask for Wayland, let's assume it can handle the standard Wayland behavior in SDL3 */
```

Comment:
This will also get hit if the user set SDL_VIDEODRIVER=wayland in the environment. In this case we definitely want to continue to set `SDL_HINT_VIDEO_WAYLAND_SCALE_TO_DISPLAY`

Reply:
Okay, this makes the problem a lot more complicated. The big problem is trying to distinguish between a user setting SDL_VIDEODRIVER=wayland and the app internally doing it.

The only way I can think to do it is to snapshot the environment with `SDL_CreateEnvironment(true)` in `SDL2Compat_InitOnStartupInternal()`, unset the variable in the real environment, and then see if we observe it set again later on. The problem with this approach is that unsetting `SDL_VIDEODRIVER` is observable by the app and could break its logic.

Maybe a quirk is best after all...

Reply:
In any case the cleanup seems like a good change.

Reply:
I put up a second PR #346 with just the init refactoring (and a little more than I did here)